src/human_encoding/parse/ast.rs

```diff
@@ -0,0 +1,846 @@
+
+        // Comments (single-line comments only).
+        "DEFAULT" | "LINE_COMMENT" = pattern r"--.*\n" => |lexer| lexer.skip();
+
+        // No whitespace is significant except to separate other tokens.
+        "DEFAULT" | "WS" = pattern r"\s" => |lexer| lexer.skip();
+    )
+}
+
+fn grammar<J: Jet + 'static>() -> Grammar<Ast<J>> {
+    santiago::grammar!(
+        "program" => empty => |_| Ast::Program(vec![]);
+        "program" => rules "line" "program" => |mut toks| {
+            if let Some(e) = propagate_errors(&toks) { return e; }
+            let line = toks[0].expect_line();
+            let prog = toks[1].expect_program();
+
+            let mut new_prog = Vec::with_capacity(1 + prog.len());
+            new_prog.push(line);
+            new_prog.extend(prog);
+            Ast::Program(new_prog)
+        };
+
+        "line" => rules "symbol" ":" "arrow" => |mut toks| Ast::from_2(
+            &mut toks,
+            Ast::expect_symbol,
+            Ast::expect_arrow,
+            |symb, arrow| Ast::Line(Line {
+                position: symb.1,
+                name: symb.0,
+                expression: None,
+                arrow,
+            })
+        );
+        "line" => rules "symbol" ":=" "expr" => |mut toks| Ast::from_2(
+            &mut toks,
+            Ast::expect_symbol,
+            Ast::expect_expression,
+            |symb, expr| Ast::Line(Line {
+                position: symb.1,
+                name: symb.0,
+                expression: Some(expr),
+                arrow: (None, None),
+            })
+        );
+        "line" => rules "symbol" ":=" "expr" ":" "arrow" => |mut toks| Ast::from_3(
+            &mut toks,
+            Ast::expect_symbol,
+            Ast::expect_expression,
+            Ast::expect_arrow,
+            |symb, expr, arrow| Ast::Line(Line {
+                position: symb.1,
+                name: symb.0,
+                expression: Some(expr),
+                arrow,
+            }),
+        );
+
+        "arrow" => rules "type" "->" "type" => |mut toks| Ast::from_2(
+            &mut toks,
+            Ast::expect_type,
+            Ast::expect_type,
+            Ast::Arrow,
+        );
+
+        "expr" => rules "symbol" => |mut toks| {
+            if let Some(e) = propagate_errors(&toks) { return e; }
+            assert_eq!(toks.len(), 1);
+            if let Ast::Symbol { value, position } = mem::replace(&mut toks[0], Ast::Replaced) {
+                Ast::Expression(Expression::reference(value, position))
+            } else {
+                unreachable!("expected string, got something else")
+            }
+        };
+        "expr" => rules "(" "expr" ")" => |toks| toks[1].clone();
+        "expr" => rules "nullary" => Ast::from_combinator;
+        "expr" => rules "unary" "expr" => Ast::from_combinator;
+        "expr" => rules "binary" "expr" "expr" => Ast::from_combinator;
+
+        // TODO should we allow CMRs as literals for constant words?
```

Comment:
8e711fd6b8a34c454d7ec83a8ad672cc6a8e12f1: What does this todo mean? Do CMRs (`#...` / `#{...}`) have a role outside of assertions?

Reply:
@uncomputable yes, they are provided as input to the first child of the `disconnect` combinator.